time.js

```diff
@@ -90,13 +90,13 @@ function _parseISO8601 (isoStr) {
     LOCAL_DATE: /^(-?(?:[1-9][0-9]*)?[0-9]{4})-(1[0-2]|0[1-9])-(3[01]|0[1-9]|[12][0-9])$/, // YYYY-MM-DD
     LOCAL_TIME: /^\d{2}:\d{2}(:\d{2})?(\.\d+)?$/, // hh:mm or hh:mm:ss or hh:mm:ss.f
     LOCAL_DATE_TIME: /^(-?(?:[1-9][0-9]*)?[0-9]{4})-(1[0-2]|0[1-9])-(3[01]|0[1-9]|[12][0-9])T\d{2}:\d{2}(:\d{2})?(\.\d+)?$/, // LOCAL_DATE and LOCAL_TIME connected with "T"
-    ISO_8160_FULL: /^(-?(?:[1-9][0-9]*)?[0-9]{4})-(1[0-2]|0[1-9])-(3[01]|0[1-9]|[12][0-9])T(2[0-3]|[01][0-9]):([0-5][0-9])(:[0-5][0-9])?(\.\d+)?(Z|[+-]\d{2}(:\d{2})?)/
+    ISO_8601_FULL: /^(-?(?:[1-9][0-9]*)?[0-9]{4})-(1[0-2]|0[1-9])-(3[01]|0[1-9]|[12][0-9])T(2[0-3]|[01][0-9]):([0-5][0-9])(:[0-5][0-9])?(\.\d+)?(Z|[+-]\d{2}(:\d{2})?)/
   };
   switch (true) {
     case REGEX.LOCAL_DATE.test(isoStr): return time.ZonedDateTime.of(time.LocalDate.parse(isoStr), time.LocalTime.MIDNIGHT, time.ZoneId.SYSTEM);
     case REGEX.LOCAL_TIME.test(isoStr): return time.ZonedDateTime.of(time.LocalDate.now(), time.LocalTime.parse(isoStr), time.ZoneId.SYSTEM);
     case REGEX.LOCAL_DATE_TIME.test(isoStr): return time.ZonedDateTime.of(time.LocalDateTime.parse(isoStr), time.ZoneId.SYSTEM);
-    case REGEX.ISO_8160_FULL.test(isoStr): return time.ZonedDateTime.parse(isoStr);
+    case REGEX.ISO_8601_FULL.test(isoStr): return time.ZonedDateTime.parse(isoStr).withZoneSameLocal(time.ZoneId.SYSTEM);
```

Comment:
One thing that just came to my mind before merging, but after approving:

What if a user passes an ISO String containing a timezone? In this case his passed in timezone would be overwritten.

We’d have to only set the zone when no zone is passed in. So either we make two regexes out of the single ISO8601 and differ between with and without timezone, or we append a `[SYSTEM]` of the string does not include a timezone.

Reply:
This REGEX should only match ISO8601 with timezone in square brackets:

`^(-?(?:[1-9][0-9]*)?[0-9]{4})-(1[0-2]|0[1-9])-(3[01]|0[1-9]|[12][0-9])T(2[0-3]|[01][0-9]):([0-5][0-9])(:[0-5][0-9])?(\.\d+)?(Z|[+-]\d{2}(:\d{2})?\[[\w/]*\])`

And this one should only match ISO8601 without timezone in square brackets:

`^(-?(?:[1-9][0-9]*)?[0-9]{4})-(1[0-2]|0[1-9])-(3[01]|0[1-9]|[12][0-9])T(2[0-3]|[01][0-9]):([0-5][0-9])(:[0-5][0-9])?(\.\d+)?(Z|[+-]\d{2}(:\d{2})$)`

FYI: I usually try this out on https://regex101.com

Reply:
Eeek,  that's right. We probably don't want to just replace the timezone ID if it's used in the String, only when it's not there should we assume the system default.

> We’d have to only set the zone when no zone is passed in. So either we make two regexes out of the single ISO8601 and differ between with and without timezone, or we append a [SYSTEM] of the string does not include a timezone.

There's a third option which could make the regex simpler. We can parse the ISO8601 string as is. Then we can get the zone ID from the new ZDT. If that ID is `[+-]\d{2}(:\d{2}` we need to set the zone id to the system default. 

I don't know if that makes a difference from a maintainability perspective.

Reply:
IMO having two different regexes is more straight forward, than checking the zone id of the ZDT.

Reply:
Added a second regex and it works for Strings.

However, it still fails to work for Items and rawState. See below.

Reply:
See my comment https://github.com/openhab/openhab-js/pull/307#discussion_r1395564442.